tensor/iterator.go

```diff
@@ -172,6 +368,73 @@ func (it *FlatIterator) Chan() (retVal chan int) {
 	return
 }
 
+/* FLAT MASKED ITERATOR */
+
+// FlatMaskedIterator is an iterator that iterates over simple masked Tensors.
+// It is used when the mask stride is identical to data stride with the exception of trailing zeros,
+// in which case the data index is always a perfect integer multiple of the mask index
+type FlatMaskedIterator struct {
+	*FlatIterator
+	mask []bool
+}
+
+// NewFlatMaskedIterator creates a new flat masked iterator
+func NewFlatMaskedIterator(ap *AP, mask []bool) *FlatMaskedIterator {
+	it := new(FlatMaskedIterator)
+	it.FlatIterator = NewFlatIterator(ap)
+	it.mask = mask
+	return it
+}
+
+// FlatMaskedIteratorFromDense creates a new FlatMaskedIterator from dense tensor
+func FlatMaskedIteratorFromDense(tt *Dense) *FlatMaskedIterator {
+	it := new(FlatMaskedIterator)
+	it.FlatIterator = FlatIteratorFromDense(tt)
+	it.mask = tt.mask
```

Comment:
add `runtime.SetFinalizer` here instead.

If you set the finalizer here, you don't need to set the finalizer everywhere else. When the iterator falls out of reference (no other object holds a reference to the iterator anymore, and the function in which the iterator was created has fallen out of scope), the finalizer `destroyIterator` will be called. 

Let the runtime handle this, instead of manually managing it - I can't seem to find anywhere where you create an iterator and don't destroy it.